src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,247 @@
+use const CURLINFO_STARTTRANSFER_TIME;
+use const CURLINFO_TOTAL_TIME;
+
+final class CurlGetinfoFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	/** @var array<int,Type>|null */
+	private array $componentTypesPairedConstants;
+
+	/** @var array<string,Type>|null */
+	private array $componentTypesPairedStrings;
+
+	private Type $allComponentsTogetherType;
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'curl_getinfo';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (count($functionCall->getArgs()) < 1) {
+			return ParametersAcceptorSelector::selectSingle(
+				$functionReflection->getVariants(),
+			)->getReturnType();
+		}
+
+		$this->cacheReturnTypes();
+
+		if (count($functionCall->getArgs()) > 1) {
+			$componentType = $scope->getType($functionCall->getArgs()[1]->value);
+
+			if (!$componentType instanceof ConstantType) {
+				return $this->createAllComponentsReturnType();
+			}
+
+			$componentType = $componentType->toInteger();
+			if (!$componentType instanceof ConstantIntegerType) {
+				throw new ShouldNotHappenException();
+			}
+		} else {
+			$componentType = new ConstantIntegerType(-1);
+		}
+
+		if ($componentType->getValue() === -1) {
+			return $this->createAllComponentsReturnType();
+		}
+
+		return $this->componentTypesPairedConstants[$componentType->getValue()] ?? new ConstantBooleanType(false);
+	}
+
+	/**
+	 * @throws ShouldNotHappenException
+	 */
+	private function createAllComponentsReturnType(): Type
+	{
+		if (isset($this->allComponentsTogetherType) === false) {
+			$returnTypes = [
+				new ConstantBooleanType(false),
+			];
+
+			$builder = ConstantArrayTypeBuilder::createEmpty();
+
+			if (isset($this->componentTypesPairedStrings) === false) {
+				throw new ShouldNotHappenException();
+			}
+
+			foreach ($this->componentTypesPairedStrings as $componentName => $componentValueType) {
+				$builder->setOffsetValueType(new ConstantStringType($componentName), $componentValueType, true);
+			}
+
+			$returnTypes[] = $builder->getArray();
+
+			$this->allComponentsTogetherType = TypeCombinator::union(...$returnTypes);
+		}
+
+		return $this->allComponentsTogetherType;
+	}
+
+	private function cacheReturnTypes(): void
+	{
+		if (isset($this->componentTypesPairedConstants) === true) {
+			return;
+		}
+
+		$string = new StringType();
+		$integer = new IntegerType();
+		$float = new FloatType();
+		$false = new ConstantBooleanType(false);
+		$null = new NullType();
+
+		$stringOrNull = TypeCombinator::union($string, $null);
+		$stringOrFalse = TypeCombinator::union($string, $false);
+		$arrayOfIntegerKeyedStrings = new ArrayType($integer, $string);
+		$arrayOfArrayIntegerKeyedStrings = new ArrayType($integer, $arrayOfIntegerKeyedStrings);
+
+		$this->componentTypesPairedConstants = [
```

Comment:
Is this caching even necessary..? Do you have evidence it actually prevent a bottleneck?

Reply:
I might have used another class as inspiration, so I thought it was required. I removed the caching in a new commit.